src/Microsoft.TestPlatform.CrossPlatEngine/Client/Parallel/ParallelProxyDiscoveryManager.cs

```diff
@@ -182,21 +185,45 @@ public bool HandlePartialDiscoveryComplete(IProxyDiscoveryManager proxyDiscovery
 
     private List<ProviderSpecificWorkload<DiscoveryCriteria>> SplitToWorkloads(DiscoveryCriteria discoveryCriteria, Dictionary<string, TestRuntimeProviderInfo> sourceToTestHostProviderMap)
     {
+        var sources = discoveryCriteria.Sources;
+        // Each source is grouped with its respective provider.
+        var providerGroups = sources
+            .Select(source => new ProviderSpecificWorkload<string>(source, sourceToTestHostProviderMap[source]))
+            .GroupBy(psw => psw.Provider);
+
         List<ProviderSpecificWorkload<DiscoveryCriteria>> workloads = new();
-        foreach (var source in discoveryCriteria.Sources)
+        foreach (var group in providerGroups)
         {
-            var testHostProviderInfo = sourceToTestHostProviderMap[source];
-            var runsettingsXml = testHostProviderInfo.RunSettings;
-            var updatedDiscoveryCriteria = new ProviderSpecificWorkload<DiscoveryCriteria>(NewDiscoveryCriteriaFromSourceAndSettings(source, discoveryCriteria, runsettingsXml), testHostProviderInfo);
-            workloads.Add(updatedDiscoveryCriteria);
+            var testhostProviderInfo = group.Key;
+            // If the run is not parallel and the host is shared, put all testcases on single testhost.
+            if (!_isParallel && testhostProviderInfo.Shared)
+            {
+                var runsettings = testhostProviderInfo.RunSettings;
+                var sourcesToDiscover = group.Select(w => w.Work).ToArray();
```

Comment:
If you change `NewDiscoveryCriteriaFromSourceAndSettings` to take `IEnumerable<string>` you won't need the `ToArray` call here.

Reply:
But I want to see them when I debug. not somewhere x layers deeper, just to avoid calling toArray on collection of <10 items. Is there any perf improvement you can show me? :)

Reply:
You can call `ToList`/`ToArray` or even simply expand when debugging. I don't have any perf but the global reasoning of "it's only adding a small amount" everywhere in the code might cause problems in the long run.

Your call to fix or close :)

Reply:
Changed it anyway. If that makes you happy :D